### PR TITLE
structure path handling ( #15 )

### PR DESCRIPTION
**Pull request by danielknobe**
- use PrefDir of physfs as write folder (on windows too!)
- support XDG directory specificiation
- remove deprecated GAMEDATADIR
- sort pathes and add comments to make clear the idea behind all that

### CODE REVIEW

**Comment by ngc92** [comment]
First question, now that we are restructuring this: Do we want to keep the option for reading from a bunch of zip files? Compared to either reading from files directly, or reading from a single zip with multiple directories?

Second, in my short experimentations with this, I came to the conclusion that `probeDir` is probably broken, in that if the directory "exists", it won't create a new directory. That means that e.g. if there is a `backgrounds` directory already in `/share/blobby`, we won't generate one in the pref dir. Is that what we want?

Finally, do we know whether we get usful prefDir results also for the other systems? The documentation mentions at least OS X, so can we use this too in the `#if __APPLE__` path?

**Comment by danielknobe** [comment]
I think reading from a bunch of zip files instead of a single zip has the advantage that 
backgrounds and mods can be delivered fine-grained. Not to zip the files may be a option.
Advantage zipping files:
  - memory size (1,5MB)

Advantage NOT zipping files:
  - simplification of build process

Do you see any advantages/disadvantages?
Do you think we should change that? 

You are right with the probeDir issue. 
If we handle the writeDir (and probing the folders) first the issue is solved.
But this is not very elegant but works.

MacOS is already using prefDir in this PR.

I tested prefDir & baseDir on on iOS. I got the following.

Simulator:
  - prefDir: /Users/danielknobe/Library/Developer/CoreSimulator/Devices/2F4C9AF3-E95E-4916-B251-277602E3233F/data/Containers/Data/Application/2386E35A-DE03-4A1D-850A-D091875942F8/Library/Application Support/BlobbyVolley2/
  - baseDir: /Users/danielknobe/Library/Developer/CoreSimulator/Devices/2F4C9AF3-E95E-4916-B251-277602E3233F/data/Containers/Bundle/Application/8A024183-A04F-4815-8910-3C6307DF9310/blobby.app/

Target:
  - prefDir: /var/mobile/Containers/Data/Application/04FA8F5F-5F11-4C89-9148-AC0E96A0284F/Library/Application Support/BlobbyVolley2/
  - baseDir: /private/var/containers/Bundle/Application/7FAAAE17-37EF-46C5-A759-0E99C51BBCE6/blobby.app/

Reading the file system basics
https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html
the prefDir seems to be a good solution also on ios in my opition. I think we should write finally in the Documents folder there.

**Comment by danielknobe** [comment]
Is this first part of the changes ready for merge?

File: src/FileSystem.h [hunk 1]
```diff
@@ -80,4 +80,8 @@ class FileSystem : public ObjectCounter<FileSystem>
 
 		/// \todo ideally, this method would never be needed by client code!!
 		std::string getUserDir();
+
+		/// \brief Get the user-and-app-specific path where files can be written.
+		/// \todo ideally, this method would never be needed by client code!!
```
Comment (ngc92) [inline src/FileSystem.h:85]: does this TODO make sense? I'm not really sure what the aim is, here

File: src/main.cpp [hunk 2]
```diff
@@ -89,88 +89,80 @@ void setupPHYSFS()
...
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby");
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby/gfx.zip");
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby/sounds.zip");
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby/scripts.zip");
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby/backgrounds.zip");
+		fs.addToSearchPath(BLOBBY_INSTALL_PREFIX  "/share/blobby/rules.zip");
+	#endif
 
+	/// \todo the following lines must be refactored!!!!
+	#if (!defined __ANDROID__) && !(defined __APPLE__ && TARGET_OS_IPHONE) && (!BLOBBY_ON_DESKTOP)
```
Comment (ngc92) [inline src/main.cpp:146]: this condition is somewhat hard to understand. If we're not on android, desktop, or ios, does that leave anything but the switch?

File: src/FileSystem.h (hunk 1 above)
Comment (danielknobe) [reply src/FileSystem.h:85]: It is a copy&paste todo from method above. I'm not really sure, too. We should drop it.

File: src/main.cpp (hunk 2 above)
Comment (danielknobe) [reply src/main.cpp:146]: In the best case this part is completely obsolet.
It is only there for default write dir handling previously used (which includes switch).
I added an additional branch which changes default and switch write dir behavior.
@SilentFlyBy : Can you please check if **fix/path-handling-switch** is working for switch?

Comment (danielknobe) [reply src/main.cpp]: I merged **fix/path-handling-switch** into this PR. If we get any problems with switch we can fix it later.

File: src/main.cpp [hunk 3]
```diff
@@ -82,94 +76,47 @@ void deinit()
...
 	#if BLOBBY_ON_DESKTOP
-		std::string baseSearchPath("data" + separator);
-	#elif (defined __ANDROID__)
-		std::string baseSearchPath(SDL_AndroidGetExternalStoragePath() + separator);
-	#elif (defined __APPLE__) || (defined __SWITCH__)
-		// iOS
-		std::string baseSearchPath(PHYSFS_getBaseDir());
+		std::string baseSearchPath("data" + fs.getDirSeparator());
+	#else
+		std::string baseSearchPath(fs.getBaseDir());
```
Comment (ngc92) [inline src/main.cpp:85]: is this guaranteed to return a  path with dir separator at the end? Couldn't find anything on this in the docs, but if so, maybe put a comment here to explain the difference with the call aboce

File: src/main.cpp [hunk 4]
```diff
@@ -82,94 +76,47 @@ void deinit()
...
 	#endif
 
+	/*
+	set write dir
+	*/
+	std::string writeDir = fs.getPrefDir();
+	fs.setWriteDir(writeDir);
+	fs.probeDir("replays");
```
Comment (ngc92) [inline src/main.cpp:93]: should we change the name of  `probeDir`? I don't find it very descriptive. Maybe `ensureDir`?

**Comment by ngc92** [comment]
Now that you've simplified it this much, I think we can go even further :)
I've tried a suggestions in #115.

In all, I think we should merge this, and if things break on one of the less well tested platforms, we'll fix them.

File: src/main.cpp (hunk 3 above)
Comment (danielknobe) [reply src/main.cpp:85]: To make sure we have no issues in future I added a simple join method.

File: src/main.cpp (hunk 4 above)
Comment (danielknobe) [reply src/main.cpp:93]: I have no opinion on this ;). I think this is something we can change.

File: src/main.cpp (hunk 3 above)
Comment (ngc92) [reply src/main.cpp:85]: great idea

**Pull request merged**